src/cupbearer/detectors/abstraction/abstraction.py

```diff
@@ -1,198 +1,231 @@
 
-    def visit(node):
-        if node in visited:
-            return False, None
-        visited.add(node)
-        current_path.add(node)
-        for parent in graph[node]:
-            if parent in current_path:
-                return True, current_path
-            found_cycle, cycle = visit(parent)
-            if found_cycle:
-                return True, cycle
-        current_path.remove(node)
-        return False, None
 
-    for vertex in graph:
-        found_cycle, cycle = visit(vertex)
-        if found_cycle:
-            raise ValueError(f"Cycle found in activation graph: {cycle}")
+class _ModuleDict(nn.Module):
+    """A ModuleDict that allows '.' in keys (but not '/').
 
+    Pytorch's ModuleDict doesn't allow '.' in keys, but our activation names contain
+    a lot of periods.
+    """
 
-class Abstraction(nn.Module):
-    # TODO: I think we should likely get rid of get_default and instead just have some
-    # informal collection of helper functions for building reasonable abstractions.
-    @classmethod
-    @abstractmethod
-    def get_default(cls, model: HookedModel, size_reduction: int) -> Abstraction:
-        pass
+    def __init__(self, modules: dict[str, nn.Module]):
+        super().__init__()
+        for name in modules:
+            if "/" in name:
+                raise ValueError(
+                    f"For technical reasons, names cant't contain '/', got {name}"
+                )
+        # Pytorch's ModuleDict doesn't allow '.' in keys, so we replace them with '/'
+        modules = {name.replace(".", "/"): module for name, module in modules.items()}
+        self.dict = nn.ModuleDict(modules)
+
+    def __getitem__(self, key: str) -> nn.Module:
+        return self.dict[key.replace(".", "/")]
+
+    def __len__(self) -> int:
+        return len(self.dict)
+
+    def __iter__(self):
+        for key in self.dict:
+            yield key.replace("/", ".")
+
+    def __contains__(self, key: str) -> bool:
+        return key.replace(".", "/") in self.dict
+
+    def items(self):
+        for key, value in self.dict.items():
+            yield key.replace("/", "."), value
+
+    def values(self):
+        return self.dict.values()
+
+    def keys(self):
+        # HACK: we want to return an actual dict_keys object to make sure that
+        # e.g. equality with other dict_keys objects works as expected. So we create
+        # a dummy dictionary and return its keys.
+        # BUG: This probably still doesn't behave like a real dict_keys object.
+        # For example, if the underlying dictionary changes, this won't reflect that.
+        return {key.replace("/", "."): None for key in self.dict}.keys()
+
+
+def l2_loss(input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+    assert input.shape == target.shape
+    input = input.reshape(input.size(0), -1)
+    target = target.reshape(target.size(0), -1)
+    # Not totally clear whether mean or sum makes more sense here.
```

Comment:
I don't think that's a super strong reason tbh, but I do agree mean is probably better, since if we aggregate over different activations with different hidden dims, we probably don't want larger activations to dominate. But I think it's not clear whether just taking a mean here is a principled way of solving that